File: 1964. Find the Longest Valid Obstacle Course at Each Position/Solution.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        int n = obstacles.size(), l = 0;
        vector<int> res(n, 1), heights(n, 0);
        for (int i = 0; i < n; i++) {
            int idx = search(heights, obstacles[i], l);
            if (idx == l) l++;
            heights[idx] = obstacles[i];
            res[i] = idx + 1;
        }
        return res;
    }

    int search(vector<int>& A, int target, int right) {
        int left = 0;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (A[mid] <= target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
};

File: 1964. Find the Longest Valid Obstacle Course at Each Position/Solution.py
class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles: List[int]) -> List[int]:
        n, l = len(obstacles), 0
        res, heights = [1 for i in range(n)], []
        for i, h in enumerate(obstacles):
            idx = bisect.bisect_right(heights, h)
            if idx == len(heights):
                heights.append(h)
            else:
                heights[idx] = h
            res[i] = idx + 1
        return res